contracts/chain-adapters/Arbitrum_CustomGasToken_Adapter.sol

```diff
@@ -0,0 +1,291 @@
+                CUSTOM_GAS_TOKEN.safeIncreaseAllowance(address(L1_INBOX), amountToBridge);
+                L1_INBOX.createRetryableTicket(
+                    to, // destAddr destination L2 contract address
+                    L2_CALL_VALUE, // l2CallValue call value for retryable L2 message
+                    L2_MAX_SUBMISSION_COST, // maxSubmissionCost Max gas deducted from user's L2 balance to cover base fee
+                    L2_REFUND_L2_ADDRESS, // excessFeeRefundAddress maxgas * gasprice - execution cost gets credited here on L2
+                    L2_REFUND_L2_ADDRESS, // callValueRefundAddress l2Callvalue gets credited here on L2 if retryable txn times out or gets cancelled
+                    RELAY_MESSAGE_L2_GAS_LIMIT, // maxGas Max gas deducted from user's L2 balance to cover L2 execution
+                    L2_GAS_PRICE, // gasPriceBid price bid for L2 execution
+                    amountToBridge, // tokenTotalFeeAmount amount of fees to be deposited in native token.
+                    "0x" // data ABI encoded data of L2 message
+                );
+            } else {
+                CUSTOM_GAS_TOKEN.safeIncreaseAllowance(erc20Gateway, amount);
+
+                // To pay for gateway outbound transfer with custom gas token, encode the tokenTotalFeeAmount in the data field:
+                // The data format should be (uint256 maxSubmissionCost, bytes extraData, uint256 tokenTotalFeeAmount).
+                // Source: https://github.com/OffchainLabs/token-bridge-contracts/blob/5bdf33259d2d9ae52ddc69bc5a9cbc558c4c40c7/contracts/tokenbridge/ethereum/gateway/L1OrbitERC20Gateway.sol#L57
+                bytes memory data = abi.encode(L2_MAX_SUBMISSION_COST, "", requiredL1TokenTotalFeeAmount);
+                L1_ERC20_GATEWAY_ROUTER.outboundTransferCustomRefund(
+                    l1Token,
+                    L2_REFUND_L2_ADDRESS,
+                    to,
+                    amount,
+                    RELAY_TOKENS_L2_GAS_LIMIT,
+                    L2_GAS_PRICE,
+                    data
+                );
+            }
+        }
+        emit TokensRelayed(l1Token, l2Token, amount, to);
+    }
+
+    /**
+     * @notice Returns required amount of gas token to send a message via the Inbox.
+     * @return amount of gas token that this contract needs to hold in order for relayMessage to succeed.
+     */
+    function getL1CallValue(uint32 l2GasLimit) public view returns (uint256) {
+        return L2_MAX_SUBMISSION_COST + L2_GAS_PRICE * l2GasLimit;
+    }
+
+    function _pullCustomGas(uint32 l2GasLimit) internal returns (uint256) {
+        uint256 requiredL1CallValue = getL1CallValue(l2GasLimit);
+        CUSTOM_GAS_TOKEN_FUNDER.withdraw(CUSTOM_GAS_TOKEN, requiredL1CallValue);
+        require(CUSTOM_GAS_TOKEN.balanceOf(address(this)) >= requiredL1CallValue, "Insufficient gas balance");
+        return requiredL1CallValue;
+    }
```

Comment:
Just to be really explicit: the reason we don't want to just pull these tokens from msg.sender is that these executions are often called via the external multicall contract, so this would mean that we could no longer batch these executions with the mainnet spoke pool relayer refund leaf execution, right?

Reply:
> we could no longer batch these executions with the mainnet spoke pool relayer refund leaf execution

Why not?

Reply:
Sorry, I don't think I said that very clearly. We can batch with this implementation.

If we didn't do the funder contract and just did `transferFrom(msg.sender, ...)`, that would work, but it would break batching since we use the external multicall contract to do batching now, and that contract would be the msg.sender (and contains no tokens).